### ISSUE TITLE
stagedfunction fails to update cache

### ISSUE THREAD

**Issue by mauro3** [JuliaLang/julia#10178]
In some corner cases it seems that the stagedfunction cache is not updated.  Consider:

``` julia
immutable AI end

###############
print("\n\nThis works only in older julia-0.4\n")

print("\n1\n")
stagedfunction f{X}(::Type{AI}, x::X)
    print("\nstaging\n")
    :(x)
end

g(x) = f(AI, x)
@show g(5)

print("\n2\n")
stagedfunction f{X}(::Type{AI}, x::X)
    print("\nstaging\n")
    :(2x)
end

g(x) = f(AI, x)
@show g(5)


print("\n3\n")
stagedfunction f{X}(::Type{AI}, x::X)
    print("\nstaging\n")
    :(3x)
end

g(x) = f(AI, x)
@show g(5)

###############
print("\n\nWorks in both\n")

print("\n1\n")
stagedfunction ff{X}(::Type{AI}, x::X)
    print("\nstaging\n")
    :(x)
end
@show ff(AI, 5)

print("\n2\n")
stagedfunction ff{X}(::Type{AI}, x::X)
    print("\nstaging\n")
    :(2x)
end

@show ff(AI, 5)


print("\n3\n")
stagedfunction ff{X}(::Type{AI}, x::X)
    print("\nstaging\n")
    :(3x)
end

@show ff(AI, 5)
```

With version  bcd0e5e of julia-0.4 (mid-January) I get output:

```
This works only in older julia-0.4

1

staging

staging
g(5) = 5

2

staging

staging
g(5) = 10

3

staging

staging
g(5) = 15


Works in both

1

staging
ff(AI,5) = 5

2

staging
ff(AI,5) = 10

3

staging
ff(AI,5) = 15
```

with a current build 1a9e1b14ba I get this:

```

This works only in older julia-0.4

1

staging
g(5) = 5

2
g(5) = 5

3
g(5) = 5


Works in both

1

staging
ff(AI,5) = 5

2

staging
ff(AI,5) = 10

3

staging
ff(AI,5) = 15
```

I suspect the culprit is PR  #9921

**Comment by mauro3** [comment]
Sorry, this could have been more succinct:
```julia
    stagedfunction f{X}(x::X)
        :(x)
    end
    g(x) = f(x)
    @assert g(5)==5
    stagedfunction f{X}(x::X)
        :(2x)
    end
    g(x) = f(x)
    @assert g(5)==10
```
Throws in current 0.4 but worked in bcd0e5e.  Bonus trick, works in both versions if wrapped in a let-block.

**Comment by mauro3** [comment]
And last bit of noise: @Jutho

**Comment by Jutho** [comment]
Well, this is clearly caused by the `stagedcache` introduced in my PR #9921, since overwriting it with the old definition by typing this in the REPL
```julia
    function Base.func_for_method(m::Method, tt, env)
        if !m.isstaged
            return m.func.code
        else
            f=ccall(:jl_instantiate_staged,Any,(Any,Any,Any),m,tt,env)
            return f.code
        end
    end
```
fixes your problem. But then again, we knew this could happen if PR #9921 got merged without anybody looking at it who actually understands what's going on in `inference.jl`. Maybe it is time to introduce a `stagedfunction` label for grouping all the remaining issues with `stagedfunction`s?

**Comment by mauro3** [comment]
Thanks for work-around.  And a stagedfuntions label would probably be good.

**Comment by mauro3** [comment]
This was resolved in julia-0.5.  Is the test as in https://github.com/JuliaLang/julia/issues/10178#issuecomment-74101117 desired?
